### ISSUE TITLE
Snippets Not Displaying Unless Browser is Restarted

### ISSUE THREAD

**Issue by stsegovia-moz** [mozmeao/snippets-service#918]
Hi Giorgos, 

TL; DR: We are observing issues where snippets do not display on the browser unless you restart the browser. 

I have already detailed this problem via email, but was able to reproduce the issue on another computer. Ben Sternthal, Senior Web Development Manager, reached out to us because he wasn't seeing the snippets when he opened up a new tab. I already knew the solution would be to restart the Firefox browser, but first we made sure to grab screenshots. 

At 11:16AM PST, Ben sends me a screenshot of his issue. In the screenshot you can see:

1. There is no visible snippet in the new tab
2. He has the current browser release (65.0.1)
3. He has the browser console dev tools open and there is an error displaying. 

<img width="1255" alt="screen shot 2019-02-19 at 11 16 25 am" src="https://user-images.githubusercontent.com/47153345/53048543-dd34e900-3449-11e9-9141-dffc01e9f588.png">

At 11:22AM PST, we attempt to use the Snippet Dev Tools `about:newtab#asrouter`. We filter by Snippets and see that there are snippets that are not blocked and should be displaying.

<img width="1102" alt="screen shot 2019-02-19 at 11 22 26 am" src="https://user-images.githubusercontent.com/47153345/53048877-9f849000-344a-11e9-8654-45eaeec2f564.png">

During this time, Ben opened up several new tabs to be sure that it wasn't a one time mistake.

After we took screenshots, Ben closed his Firefox browser and re-opened the browser. He confirmed that he was able to see snippets again. No screenshot available for the snippets re-appearing. 

I'm not sure how to isolate the issue. But I can conclude that:
 1) Snippets are not appearing when they are supposed to and 
 2) Restarting the browser seems to fix this

Please let me know if you have questions. I will continue to document whenever this issue re-occurs. Thank you!

**Comment by glogiotatidis** [comment]
@stsegovia-moz that's another very detailed report, thanks!

I'll ping folks from Activity Stream since this appears to be on their end.

**Comment by glogiotatidis** [comment]
Hey @piatra @k88hudson there seems to be a issue with snippets not displaying. This have been reported at least twice by different users on different computers. Earlier comments on this issue include a very detailed report from @stsegovia-moz. 

Questions:
 - Does the console error provide a lead?
 - Do you have ideas on what's wrong?
 - Do you need us to collect more info?

**Comment by glogiotatidis** [comment]
Also added this issue in tomorrow's Snippets + AS mtg agenda

**Comment by piatra** [comment]
Because when the issues happens snippets are not showing up in any new (new tab) the issue has to be in the main process (also confirmed by the fact that it requires a browser restart). The errors shown in the web developer console are not relevant to this issue. 
If the issue happens again it would be useful to have a screenshot from the Browser Console. 
Tools -> Web Developer -> Browser Console

Thanks.

**Comment by piatra** [comment]
If the issue reproduces it would be better to post the current state of ASRouter.jsm from your browser.
1. In order to ensure Browser Toolbox is on. Enable these checkboxes from Developer Tools settings
<img width="473" alt="image" src="https://user-images.githubusercontent.com/810040/53424165-9e1f0e80-39da-11e9-9548-ed18a617a8ed.png">
2. Go to Browser Toolbox
<img width="257" alt="image" src="https://user-images.githubusercontent.com/810040/53424277-d292ca80-39da-11e9-97fc-569c46e8479c.png">
3. Go to the debugger pane and look for ASRouter.jsm
<img width="726" alt="image" src="https://user-images.githubusercontent.com/810040/53424319-e4746d80-39da-11e9-8700-7661aeb21173.png">
4. Set a breakpoint in the `onMessage` function
<img width="437" alt="image" src="https://user-images.githubusercontent.com/810040/53424355-efc79900-39da-11e9-9d90-faa9bee3387e.png">
5. And trigger it by opening a new tab. Paste in `copy(JSON.stringify(this.state, null, 2))` to copy the state of the component and paste it here in a gist or something similar.

**Comment by stsegovia-moz** [comment]
Hi @piatra Thank you for your help! As a non-tech person, I think I'm following how to investigate the issue. I will reach out if I need help. Thank you!

**Comment by stsegovia-moz** [comment]
Actually, @piatra I got to the last step. Could you elaborate what you mean in this last step? I'm a bit confused. I want to be prepared for when we are able to replicate this issue. Thank you!

> 5. And trigger it by opening a new tab. Paste in `copy(JSON.stringify(this.state, null, 2))` to copy the state of the component and paste it here in a gist or something similar.

**Comment by piatra** [comment]
@stsegovia-moz When the code executes it will stop/pause at the breakpoint you set (line 1058 for example, in the screenshot from step4). With the code paused you can then check the state of the program.
So after step 4 you can make the code we are interested in run by opening a new tab.
Run the instruction in step 5 to copy all the variables and state from the program and share it to help debugging.

**Comment by stsegovia-moz** [comment]
Thank you! I realize I didn't even know what I break point was. Ok, I'll try playing with this and reach out if I have any more questions.

**Comment by stsegovia-moz** [comment]
So I'm following the breaking point now. That makes senses. 

One more question: Is the "copy" code from for step 5 should be set as the conditional statement for the breaking point? If not, where should I run the instruction?

**Comment by piatra** [comment]
You should select the `Console` tab and paste it in there 
<img width="377" alt="image" src="https://user-images.githubusercontent.com/810040/53474664-9eb0b700-3a65-11e9-95f2-9808e1af0b69.png">

**Comment by glogiotatidis** [comment]
This is tracked on the AS side in https://bugzilla.mozilla.org/show_bug.cgi?id=1530714

**Comment by stsegovia-moz** [comment]
@piatra I tested that code from step 5 and pasted it into console. All set! I'm now ready to test this out and document the issue the next time we encounter it. Thanks for the thorough explanation! :)

**Comment by stsegovia-moz** [comment]
Ok, so I was finally able to reproduce the issue in my browser with screenshots. This happened to me on Saturday 3/9/19 (Saturday March 19th) early morning around 1 am.

So at 1:27am I notice I have no snippets while I was clearing out my cookies/temp files, which I believe is unrelated.

![Screen Shot 2019-03-09 at 1 27 12 AM](https://user-images.githubusercontent.com/47153345/54148414-614b1280-43f1-11e9-8838-0c9e57863143.png)
![Screen Shot 2019-03-09 at 1 32 01 AM](https://user-images.githubusercontent.com/47153345/54148415-614b1280-43f1-11e9-9fd9-ea689979053b.png)

When I followed the instructions @piatra  have me, I was able to run the Dev Tool Box, to then open a new tab and copy the code into the console. I pasted the text into a txt file here. (I hope I did this correctly)
[Snippet Issue 3_9.txt](https://github.com/mozmeao/snippets-service/files/2953718/Snippet.Issue.3_9.txt)

As you can see in this screenshot, the snippets returned when I restarted my browser. 
![Screen Shot 2019-03-09 at 1 41 25 AM](https://user-images.githubusercontent.com/47153345/54148659-d3235c00-43f1-11e9-820f-38415a1fde51.png)

**Comment by glogiotatidis** [comment]
We haven't been able to reproduce this in two months. I will keep the issue open for another month and then close it.

/remind me to check in 1 month

**Comment by glogiotatidis** [comment]
Nothing new to report. Closing
